Add shared GHA for ruby

girijaasoni opened the pull request.

archanaserver commented on Gemfile:
```diff
@@ -2,6 +2,8 @@ source "https://rubygems.org"
 
 gemspec
 
+gem 'hammer_cli_foreman', github: 'theforeman/hammer-cli-foreman', branch: 'master'
```
I wonder why do we need this here, when we already have this dependency in .gemspec file?

ofedoren commented on hammer_cli_foreman_bootdisk.gemspec:
```diff
@@ -19,6 +19,7 @@ Gem::Specification.new do |s|
   s.extra_rdoc_files = Dir['LICENSE', 'README*']
   s.require_paths = ["lib"]
 
-  s.add_dependency 'hammer_cli_foreman', '>= 0.1.2'
```
Please, don't pin hammer core version like that. We mostly use "minimal" version without any upper limit. In worst cases we limit up to next major version, e.g. 4.0.0.

ofedoren commented on Gemfile:
```diff
@@ -3,10 +3,19 @@ source "https://rubygems.org"
+  gem 'minitest'
+  gem 'minitest-spec-context'
+  gem 'mocha'
+  gem 'rake', '~> 13.0'
```
I don't think we need to pin rake. If tests are broken, we should fix that. If there's a breaking change in the lib, we should adjust.

ofedoren reviewed (commented):
I've tried to push this commit on my fork and create a PR here, it doesn't trigger the GHA as well. I think GHA workflow file should be present in master and then we can trigger it with different PR. Do I understand it right, @ekohl (sorry for pinging you again regarding GHA :/ )?

Also, hammer plugins used to be tested against latest master branch of hammer-cli[-foreman], but AFAIU shared `test-gem` workflow simply installs the gem with dependencies (such as particular/latest released version of a dependency) in .gemspec/Gemfile and triggers its tests.

I'm not sure if it's sufficient. I mean, it's great that it will run the tests against already released version, but then we won't be able to see through CI if the latest changes of unreleased version caused some regressions. The only way to know that will be old good checkout with Gemfile.local and running the tests locally.

If I was adding something similar that we have now in Foreman, I'd rather create a separate workflow like `hammer_plugin` instead of `test-gem`.

Also some comments inline.

nofaralfasi commented on Rakefile:
```diff
@@ -3,9 +3,20 @@ require 'bundler/gem_tasks'
 require 'hammer_cli/i18n/find_task'
 require_relative "./lib/hammer_cli_foreman_bootdisk/version"
 require_relative "./lib/hammer_cli_foreman_bootdisk/i18n"
+require 'ci/reporter/rake/minitest'
```
Do we need this even though there are no tests in this repo?

nofaralfasi commented on Rakefile:
```diff
@@ -3,9 +3,20 @@ require 'bundler/gem_tasks'
+
+Rake::TestTask.new do |t|
+  t.libs.push 'lib'
+  t.test_files = Dir.glob('test/**/*_test.rb')
```
Same here as above

ekohl commented:
> I've tried to push this commit on my fork and create a PR here, it doesn't trigger the GHA as well. I think GHA workflow file should be present in master and then we can trigger it with different PR. Do I understand it right, @ekohl (sorry for pinging you again regarding GHA :/ )?

I added a comment to describe it: https://community.theforeman.org/t/convert-foreman-plugins-continuous-integration-to-github-actions/35917/7

Now it may also need to be triggered once, so you may need to modify the triggers (temporarily).

ofedoren commented:
Hmm... first time I see that open-ended dependency for a gem causes fails...

ekohl commented:
Building a gem gives warnings about it and validating it makes those warnings errors. Perhaps you can teach it to be more tolerant but I generally think it's a good idea.

girijaasoni commented:
> >so you may need to modify the triggers (temporarily).

It doesn't have the test folder, so I dropped the test-gem 
updated just the rubocop for the style enforcement. Would that be enough? @ekohl @ofedoren

nofaralfasi commented on .github/workflows/ruby.yml:
```diff
@@ -0,0 +1,17 @@
+  rubocop:
+    name: Rubocop
+    uses: theforeman/actions/.github/workflows/rubocop.yml@v0
+    
```
```suggestion

```

nofaralfasi commented on .rubocop.yml:
```diff
@@ -0,0 +1,26 @@
+  Enabled: false
+
+Style/SymbolProc:
+  Enabled: false
```
Missing a new line here

nofaralfasi commented on Gemfile:
```diff
@@ -1,12 +1,19 @@
 gem 'gettext', '>= 3.1.3', '< 4.0.0'
 gem 'rake', '~> 13.1.0'
 
+gem 'hammer_cli_foreman', github: 'theforeman/hammer-cli-foreman', branch: 'master'
```
Is this necessary?

girijaasoni replied:
Not with the updated version :D

nofaralfasi commented on hammer_cli_foreman_bootdisk.gemspec:
```diff
@@ -12,13 +10,13 @@ Gem::Specification.new do |s|
   s.extra_rdoc_files = Dir['LICENSE', 'README*']
   s.require_paths = ["lib"]
 
   s.add_dependency 'hammer_cli_foreman', '>= 0.1.2'
```
```suggestion
  s.add_dependency 'hammer_cli_foreman', '~> 3.10'
```

nofaralfasi commented on .github/workflows/ruby.yml:
```diff
@@ -0,0 +1,16 @@
+jobs:
+  rubocop:
+    name: Rubocop
+    uses: theforeman/actions/.github/workflows/rubocop.yml@v0
```
Missing a new line (with no spaces).

nofaralfasi commented on .rubocop.yml:
```diff
@@ -0,0 +1,27 @@
+
+Style/SymbolProc:
+  Enabled: false
+  
```
Same here - the file should end with an empty line (no spaces).

girijaasoni committed:
Add shared GHAs for ruby
